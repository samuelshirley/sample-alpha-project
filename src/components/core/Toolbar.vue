<template>
  <v-toolbar app>
    <v-toolbar-title class="headline">
      <span>Vuetify Repo</span>
    </v-toolbar-title>
    <v-spacer />
    <v-btn
      @click="requestLogin"
      v-text="'GitHub Login'"
    />
    <v-toolbar-title>
      <span class="subheading font-weight-light">Requests Remaining: {{ remaining }}</span>
    </v-toolbar-title>
    <v-toolbar-title>
      <span class="subheading font-weight-light">Rate limit resets at: {{ resetOn }}</span>
    </v-toolbar-title>
  </v-toolbar>
</template>

<script>
export default {
  name: 'Toolbar',
  data () {
    return {
      remaining: 60, // default for unauthenticated requests
      resetOn: undefined,
      intervalId: 0,
    }
  },
  created () {
    this.getLimitInfo()
    this.intervalId = setInterval(this.getLimitInfo, 30000)
  },

  destroyed () {
    clearInterval(this.intervalId)
  },

  methods: {
    async getLimitInfo () {
      const previous = this.remaining
      const { data } = await this.$http.get('https://api.github.com/rate_limit')
      this.remaining = data.resources.core.remaining
      // reset time is given in seconds, convert to ms to pass to Date obj
      this.resetOn = new Date(data.resources.core.reset * 1000).toLocaleString()
      if (this.remaining === 0 && previous !== 0) {
        this.$snackbar.error('Rate limit is reached. Try again after reset.')
      }
    },
    async requestLogin () {
      const code = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
      localStorage.setItem('code', code)

      try {
        window.open(`https://github.com/login/oauth/authorize?client_id=${process.env.VUE_APP_CLIENT_ID}&redirect_uri=http://localhost:8080/auth&state${code}`, '_self')
      } catch (err) {
        console.log(err.response)
      }
    },
  },
}
</script>
