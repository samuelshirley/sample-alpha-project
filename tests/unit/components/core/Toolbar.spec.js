import { mount } from '@vue/test-utils'

import Toolbar from '@/components/core/Toolbar'

describe('Toolbar.vue', () => {
  it('should match snapshot', () => {
    const wrapper = mount(Toolbar, {
      mocks: {
        $http: {
          get: jest.fn().mockResolvedValue({ data: { resources: { core: {} } } }),
        },
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should fetch rate limit on create', () => {
    const spy = jest.fn()
    mount(Toolbar, {
      mocks: {
        $http: {
          get: spy,
        },
      },
    })
    expect(spy).toHaveBeenCalledWith('https://api.github.com/rate_limit')
  })
})
