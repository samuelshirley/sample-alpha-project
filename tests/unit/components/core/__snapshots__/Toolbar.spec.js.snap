// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar.vue should match snapshot 1`] = `
<nav class="v-toolbar v-toolbar--fixed theme--light" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
  <div class="v-toolbar__content" style="height: 64px;">
    <div class="v-toolbar__title headline"><span>Vuetify Repo</span></div>
    <div class="spacer"></div> <button type="button" class="v-btn theme--light">GitHub Login</button>
    <div class="v-toolbar__title"><span class="subheading font-weight-light">Requests Remaining: 60</span></div>
    <div class="v-toolbar__title"><span class="subheading font-weight-light">Rate limit resets at: </span></div>
  </div>
</nav>
`;
